src/System.Text.Primitives/System/Text/Parsing/InvariantSigned.cs

```diff
@@ -899,176 +899,89 @@ public unsafe static bool TryParseInt32(byte* text, int length, out int value, o
-                    // digits. There may be no overflow after Int32OverflowLength if there are leading zeroes.
-                    for (int index = indexOfFirstDigit + 1; index < overflowLength - 1; index++)
+
+                    int firstNonZeroDigitIndex = index;
+                    if (textLength - firstNonZeroDigitIndex < Int32OverflowLength)
                     {
-                        int nextDigit = text[index] - 48; // '0'
-                        if (nextDigit < 0 || nextDigit > 9)
+                        do
                         {
-                            bytesConsumed = index;
-                            value = parsedValue * sign;
-                            return true;
-                        }
-                        parsedValue = parsedValue * 10 + nextDigit;
+                            answer = answer * 10 + num - '0';
+                            index++;
+                            if (index >= textLength) goto Done;
+                            num = text[index];
+                        } while (num >= '0' && num <= '9');
                     }
-                    for (int index = overflowLength - 1; index < text.Length; index++)
+                    else
                     {
-                        int nextDigit = text[index] - 48; // '0'
-                        if (nextDigit < 0 || nextDigit > 9)
+                        do
                         {
-                            bytesConsumed = index;
-                            value = parsedValue * sign;
-                            return true;
-                        }
-                        // If parsedValue > (int.MaxValue / 10), any more appended digits will cause overflow.
-                        // if parsedValue == (int.MaxValue / 10), any nextDigit greater than 7 or 8 (depending on sign) implies overflow.
-                        bool positive = sign > 0;
-                        bool nextDigitTooLarge = nextDigit > 8 || (positive && nextDigit > 7);
-                        if (parsedValue > int.MaxValue / 10 || parsedValue == int.MaxValue / 10 && nextDigitTooLarge)
-                        {
-                            bytesConsumed = 0;
-                            value = default(int);
-                            return false;
-                        }
-                        parsedValue = parsedValue * 10 + nextDigit;
+                            answer = answer * 10 + num - '0';
+                            index++;
+                            if (index - firstNonZeroDigitIndex == Int32OverflowLength - 1)
+                            {
+                                num = text[index];
+                                if (num >= '0' && num <= '9')
+                                {
+                                    num -= (byte)'0';
+                                    if (WillOverFlow(answer, num, sign)) goto FalseExit;
```

Comment:
right, i did not realize that. Most of this solution would not be applicable then.